pkg/server/plugin/noderesolver/aws/iid.go

```diff
@@ -0,0 +1,282 @@
+			Name: aws.String("instance-state-name"),
+			Values: []*string{
+				aws.String("pending"),
+				aws.String("running"),
+			},
+		},
+	}
+)
+
+type awsClient interface {
+	DescribeInstancesWithContext(aws.Context, *ec2.DescribeInstancesInput, ...request.Option) (*ec2.DescribeInstancesOutput, error)
+	GetInstanceProfileWithContext(aws.Context, *iam.GetInstanceProfileInput, ...request.Option) (*iam.GetInstanceProfileOutput, error)
+}
+
+type IIDResolverConfig struct {
+	AccessKeyID     string `hcl:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key"`
+}
+
+type IIDResolverPlugin struct {
+	mu      sync.RWMutex
+	conf    *aws.Config
+	clients map[string]awsClient
+
+	hooks struct {
+		getenv    func(string) string
+		newClient func(conf *aws.Config) (awsClient, error)
+	}
+}
+
+var _ noderesolver.Plugin = (*IIDResolverPlugin)(nil)
+
+func NewIIDResolverPlugin() *IIDResolverPlugin {
+	p := &IIDResolverPlugin{}
+	p.hooks.getenv = os.Getenv
+	p.hooks.newClient = newAWSClient
+	return p
+}
+
+func (p *IIDResolverPlugin) Configure(ctx context.Context, req *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
+	// Parse HCL config payload into config struct
+	config := new(IIDResolverConfig)
+	if err := hcl.Decode(config, req.Configuration); err != nil {
+		return nil, iidError.New("unable to decode configuration: %v", err)
+	}
+
+	// Set defaults from the environment
+	if config.AccessKeyID == "" {
+		config.AccessKeyID = p.hooks.getenv("AWS_ACCESS_KEY_ID")
+	}
+	if config.SecretAccessKey == "" {
+		config.SecretAccessKey = p.hooks.getenv("AWS_SECRET_ACCESS_KEY")
+	}
+
+	conf := aws.NewConfig()
+	switch {
+	case config.AccessKeyID != "" && config.SecretAccessKey != "":
+		conf.Credentials = credentials.NewStaticCredentials(config.AccessKeyID, config.SecretAccessKey, "")
+	case config.AccessKeyID != "" && config.SecretAccessKey == "":
+		return nil, iidError.New("configuration missing secret access key")
+	case config.AccessKeyID == "" && config.SecretAccessKey != "":
+		return nil, iidError.New("configuration missing access key id")
+	case config.AccessKeyID == "" && config.SecretAccessKey == "":
+		return nil, iidError.New("configuration missing both access key id and secret access key")
+	}
+
+	// set the AWS configuration and reset clients
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.conf = conf
+	p.clients = make(map[string]awsClient)
+	return &spi.ConfigureResponse{}, nil
+}
+
+func (p *IIDResolverPlugin) GetPluginInfo(context.Context, *spi.GetPluginInfoRequest) (*spi.GetPluginInfoResponse, error) {
+	return &spi.GetPluginInfoResponse{}, nil
+}
+
+func (p *IIDResolverPlugin) Resolve(ctx context.Context, req *noderesolver.ResolveRequest) (*noderesolver.ResolveResponse, error) {
+	resp := &noderesolver.ResolveResponse{
+		Map: make(map[string]*common.Selectors),
+	}
+	for _, spiffeID := range req.BaseSpiffeIdList {
+		selectors, err := p.resolveSpiffeID(ctx, spiffeID)
+		if err != nil {
+			return nil, err
+		}
+		if selectors != nil {
```

Comment:
Nil is the zero value, so doing this check might be a moot point?

Reply:
for out-of-proc resolvers that is probably true due to the way the protobuf is serialized/deserialized via GRPC, however, the handler code assumes that any entry in the map has a non-nil selector message:

```
	if selectors, ok := response.Map[baseSpiffeID]; ok {
		for _, selector := range selectors.Entries {
			err := h.createNodeResolverMapEntry(ctx, baseSpiffeID, selector)
			if err != nil {
				return err
			}
		}
	}
```

I'd be ok fixing the handler and updating the resolver code to not care about inserting a nil entry into the map.